File: javascript/trees/MinHeap.js
/**
 * Min Heap
 *
 * Insert:      O(log n)
 * Extract min: O(log n)
 * Peek:        O(1)
 */

class MinHeap {
  constructor() {
    this.data = [];
  }

  insert(val) {
    this.data.push(val);
    this._bubbleUp(this.data.length - 1);
  }

  extractMin() {
    if (this.data.length === 0) {
      throw new Error('Heap is empty');
    }

    const min = this.data[0];
    const last = this.data.pop();

    if (this.data.length > 0) {
      this.data[0] = last;
      this._bubbleDown(0);
    }

    return min;
  }

  peek() {
    if (this.data.length === 0) {
      throw new Error('Heap is empty');
    }
    return this.data[0];
  }

  size() {
    return this.data.length;
  }

  isEmpty() {
    return this.data.length === 0;
  }

  _bubbleUp(idx) {
    while (idx > 0) {
      const parent = Math.floor((idx - 1) / 2);
      if (this.data[idx] < this.data[parent]) {
        [this.data[idx], this.data[parent]] = [this.data[parent], this.data[idx]];
        idx = parent;
      } else {
        break;
      }
    }
  }

  _bubbleDown(idx) {
    const n = this.data.length;
    while (true) {
      let smallest = idx;
      const left = 2 * idx + 1;
      const right = 2 * idx + 2;

      if (left < n && this.data[left] < this.data[smallest]) smallest = left;
      if (right < n && this.data[right] < this.data[smallest]) smallest = right;

      if (smallest !== idx) {
        [this.data[idx], this.data[smallest]] = [this.data[smallest], this.data[idx]];
        idx = smallest;
      } else {
        break;
      }
    }
  }
}

module.exports = MinHeap;

File: python/trees/min_heap.py
"""
Min Heap (Binary Heap)
----------------------
Insert:      O(log n)
Extract min: O(log n)
Peek min:    O(1)

Array-based implementation. Parent of index i is at (i-1)//2,
children are at 2i+1 and 2i+2.

Python has heapq in the standard library but building it from
scratch helps understand how priority queues actually work.
"""


class MinHeap:
    def __init__(self):
        self._data = []

    def insert(self, val):
        self._data.append(val)
        self._bubble_up(len(self._data) - 1)

    def extract_min(self):
        if not self._data:
            raise IndexError("extract from empty heap")

        min_val = self._data[0]
        last = self._data.pop()

        if self._data:
            self._data[0] = last
            self._bubble_down(0)

        return min_val

    def peek(self):
        if not self._data:
            raise IndexError("peek from empty heap")
        return self._data[0]

    def size(self):
        return len(self._data)

    def is_empty(self):
        return len(self._data) == 0

    def _bubble_up(self, idx):
        while idx > 0:
            parent = (idx - 1) // 2
            if self._data[idx] < self._data[parent]:
                self._data[idx], self._data[parent] = self._data[parent], self._data[idx]
                idx = parent
            else:
                break

    def _bubble_down(self, idx):
        n = len(self._data)
        while True:
            smallest = idx
            left = 2 * idx + 1
            right = 2 * idx + 2

            if left < n and self._data[left] < self._data[smallest]:
                smallest = left
            if right < n and self._data[right] < self._data[smallest]:
                smallest = right

            if smallest != idx:
                self._data[idx], self._data[smallest] = self._data[smallest], self._data[idx]
                idx = smallest
            else:
                break

    def to_list(self):
        return list(self._data)
